rudder-core/src/main/scala/com/normation/rudder/services/policies/write/Cf3PromisesFileWriterService.scala

```diff
@@ -0,0 +1,620 @@
+import org.apache.commons.io.IOUtils
+import org.apache.commons.io.FilenameUtils
+import org.joda.time.LocalTime
+import org.joda.time.LocalDate
+import org.joda.time.DateTime
+import net.liftweb.util.Helpers.tryo
+import net.liftweb.common.Full
+import net.liftweb.common.Box
+import net.liftweb.common.Failure
+import net.liftweb.common.Loggable
+import net.liftweb.common.EmptyBox
+import com.normation.rudder.services.policies.BundleOrder
+import com.normation.rudder.services.policies.nodeconfig.NodeConfigurationLogger
+
+
+/**
+ * Write promises for the set of nodes, with the given configs.
+ * Requires access to external templates files.
+ */
+trait Cf3PromisesFileWriterService {
+
+  /**
+   * Write templates for node configuration that changed since the last write.
+   *
+   */
+  def writeTemplate(rootNodeId: NodeId, nodesToWrite: Set[NodeId], allNodeConfigs: Map[NodeId, NodeConfiguration], versions: Map[NodeId, NodeConfigId]) : Box[Seq[NodeConfiguration]]
+}
+
+
+
+class Cf3PromisesFileWriterServiceImpl(
+    techniqueRepository      : TechniqueRepository
+  , pathComputer             : PathComputer
+  , licenseRepository        : LicenseRepository
+  , logNodeConfig            : NodeConfigurationLogger
+  , prepareTemplate          : PrepareTemplateVariables
+  , communityCheckPromises   : String
+  , novaCheckPromises        : String
+  , cfengineReloadPromises   : String
+) extends Cf3PromisesFileWriterService with Loggable {
+
+
+  /**
+   * Write templates for node configuration that changed since the last write.
+   *
+   */
+  override def writeTemplate(rootNodeId: NodeId, nodesToWrite: Set[NodeId], allNodeConfigs: Map[NodeId, NodeConfiguration], versions: Map[NodeId, NodeConfigId]) : Box[Seq[NodeConfiguration]] = {
+
+    val TAG_OF_RUDDER_ID = "@@RUDDER_ID@@"
+    val GENEREATED_CSV_FILENAME = "rudder_expected_reports.csv"
+
+    val newPostfix = ".new"
+    val backupPostfix = ".bkp"
+
+
+    val nodeConfigsToWrite = allNodeConfigs.filterKeys(nodesToWrite.contains(_))
+    val interestingNodeConfigs = allNodeConfigs.filterKeys(k => nodeConfigsToWrite.exists{ case(x, _) => x == k }).values.toSeq
+    val techniqueIds = interestingNodeConfigs.flatMap( _.getTechniqueIds ).toSet
+
+    //debug - but don't fails for debugging !
+    logNodeConfig.log(nodeConfigsToWrite.values.toSeq) match {
+      case eb:EmptyBox =>
+        val e = eb ?~! "Error when trying to write node configurations for debugging"
+        logger.error(e)
+        e.rootExceptionCause.foreach { ex =>
+          logger.error("Root exception cause was:", ex)
+        }
+      case _ => //nothing to do
+    }
+
+
+    /*
+     * Here come the general writing process
+     */
+
+    def sequencePar[U,T](seq:Seq[U])(f:U => Box[T]) : Box[Seq[T]] = {
+      val buf = scala.collection.mutable.Buffer[T]()
+      seq.par.foreach { u => f(u) match {
+        case e:EmptyBox => return e
+        case Full(x) => buf += x
+      } }
+      Full(buf)
+    }
```

Comment:
it was a test to see if perf improved, and well, there is more than a 50% gain. 
But I was wondering if there wasn't a risk of return escaping its scope if two thread of the foreach throws one at the same time. For now, I don't know :)